hanish/darksky.py

```diff
@@ -0,0 +1,192 @@
+            us, and si. The default is auto which selects units based on the
+            specified location or language.
+
+        Returns
+        -------
+        data: json
+            The parsed json response of the API query.
+        """
+        # Use the cached response if it's available and timeout is specified.
+        if self.cache_timeout:
+            data = self.cache.get((lat,lon))
+            if data is not None:
+                return data
+
+        # Create the query params
+        query = {
+            "lang": lang,
+            "units": units,
+        }
+
+        # Convert the exclude list into the correct format
+        if exclude is not None:
+            query["exclude"] = ",".join(exclude)
+
+        # Convert the extend bool into the correct format
+        if extend:
+            query["extend"] = "hourly"
+
+        # Perform the query to the Dark Sky API
+        # TODO: use safer method than simple string formatting
+        endpoint = "/forecast/{}/{},{}".format(self.apikey, lat, lon)
+        data = self.request(endpoint, **query)
+
+        # Cache the result and return
+        if self.cache_timeout:
+            self.cache[(lat,lon)] = data
+
+        return data
+
+    def request(self, endpoint, **query):
+        """
+        Performs a GET request by joining the provided endpoint with the API
+        url and encoding and attatching the query params provided as generic
+        keyword arguments. This function also tracks the number of calls made
+        to the API as reported by the X-Forecast-API-Calls header from the
+        server, preventing additional requests if the limit is reached.
+
+        Note: This method performs no cacheing and will always make a request
+              if under the per-day limit. This method will also raise an
+              exception for the HTTP status (e.g. 404, 403, etc.)
+
+        Parameters
+        ----------
+        endpoint: string
+            Path to the Dark Sky api resource, not the full URL
+
+        query: dict
+            Dictionary of HTTP query arguments to encode and request
+
+        Returns
+        -------
+        data: json
+            The parsed json response of the API query.
+        """
+
+        if self.limit is not None and self.n_api_calls > self.limit:
+            raise DarkSkyException(
+                "api query limit of {} requests reached".format(self.limit)
+            )
+
+        # Join the endpoint with the URL
+        url = urljoin(DARKSKY_API_URL, endpoint)
+
+        # Ensure that we'll accept compressed JSON to reduce bandwidth
+        headers = {'Accept-Encoding': 'gzip'}
+
+        # Perform the query, raising an exception for non-200 status
+        r = requests.get(url, params=query, headers=headers)
+        r.raise_for_status()
+
+        # Get the X-Forecast-API-Calls header
+        if API_CALLS_HEADER in r.headers:
+            self.n_api_calls = int(r.headers[API_CALLS_HEADER])
```

Comment:
But what if we hit the `limit` today? When the day turns over, will we make the call to allow the value to reset?